[Pham Chau Giang] iP

pcgiang opened the pull request:
### Duke

> "If you spend too much time thinking about a thing, you'll never get it done." - Bruce Lee ([source](https://dansilvestre.com/productivity-quotes/))

No longer need to remember things you need to do, Duke is:

- text-based
- super easy to learn
- ~~FAST~~ SUPER FAST to use!!

What you need to do is:
1. Download Duke from [here](https://nus-cs2103-ay2122s1.github.io/website/schedule/week4/project.html)
2. Click twice!
3. Add your tasks.
4. Trust Duke and we will manage your tasks for you! 👯 

And it is **FREE**!!

Features:

- [x] Manage tasks
- [x] Manage deadlines and time for events
- [ ] Reminders (coming soon)

If you wish to practise Java with our app, here's the our `main` method:
```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

pcgiang committed:
Level-1

pcgiang committed:
Level 2 is completed!

pcgiang committed:
Level-3 is done!

pcgiang committed:
Level-4 is done!

pcgiang committed:
A-TextUiTesting is done!!

pcgiang committed:
Level-5 is done!!

pcgiang committed:
Level 6 is completed!!

Add in delete method. Realised that the printing methods from previous levels did not function correctly. Edit print methods and EXPECTED.txt file

pcgiang committed:
Edit Exception message

pcgiang committed:
Level-7 is done!

pcgiang committed:
Level-7 is done!

pcgiang committed:
Update deadline attribute in Deadline class, create a new exception

pcgiang committed:
Merge branch 'branch-level-8'

pcgiang committed:
A-MoreOOP is done!!!

pcgiang committed:
A-Packages

pcgiang committed:
A-JUnit

pcgiang committed:
A-JavaDoc

Add Java Doc for public methods and classes.

pcgiang committed:
A-JavaDocs!!

pcgiang committed:
A-CodingStandard

pcgiang committed:
A-CodingStandard

pcgiang committed:
Merge branch 'branch-A-CodingStandard'

pcgiang committed:
Update Javadoc for TaskList

pcgiang committed:
Merge remote-tracking branch 'upstream/add-gradle-support' into A-gradle

pcgiang committed:
step 3

pcgiang committed:
JavaFX learning stuff

over-fitted commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,78 @@
+
+        while (!input.equals("bye")) {
+            if (input.equals("list")) {
+                System.out.println("Here are the tasks on your list: ");
+                list.printList();
+            } else if (input.startsWith("done")) {
+                try {
+                    list.markAsDone(input);
+                } catch (DukeException e) {
+                    ui.printErrorMessage(e.getMessage());
+                } catch (NumberFormatException e) {
+                    ui.printErrorMessage("Task number is invalid!");
+                }
+            } else if (input.startsWith("delete")) {
+                try {
+                    list.deleteTask(input);
+                } catch (DukeException e) {
+                    ui.printErrorMessage(e.getMessage());
+                }
+            } else if (input.startsWith("find")) {
+                try {
+                    list.findTask(input);
+                } catch (DukeException e) {
+                    ui.printErrorMessage(e.getMessage());
+                }
+            } else {
+                try {
+                    list.addTask(input);
+                } catch (DukeException e) {
+                    ui.printErrorMessage(e.getMessage());
+                }
+            }
+            storage.updateFile(list);
+            input = sc.nextLine();
+        }
+
+        ui.printByeMessage();
+        sc.close();
+    }
+}
```
The if and elseif chunks seem to make the code harder to read. Could the nesting be reduced, such as by using guard clauses and `break` to eliminate the need for `else`?

over-fitted commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,94 @@
+     */
+    public void updateFile(TaskList list) throws IOException {
+        File file = new File(filePath);
+        if (!file.exists()) {
+            file.createNewFile();
+        } else {
+            file.delete();
+            File newFile = new File(filePath);
+            newFile.createNewFile();
+            FileWriter writer = new FileWriter(file);
+
+            for (Task task : list.getList()) {
+                writer.write(task.writeTask() + "\n");
+            }
+            writer.close();
+        }
```
Could this block be refactored so as to remove the need for the else block and reduce indentation?
```suggestion
        if (!file.exists()) {
            file.createNewFile();
            return;
        }
        file.delete();
        File newFile = new File(filePath);
        newFile.createNewFile();
        FileWriter writer = new FileWriter(file);

        for (Task task : list.getList()) {
            writer.write(task.writeTask() + "\n");
        }
        writer.close();
```

pcgiang replied:
yep that's a good suggestion, thanks!

over-fitted commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,46 @@
+     * @return Details of task
+     */
+    public String writeTask() {
+        return (done ? "1" : "0") + " | " + title;
```
Could this be written more explicitly to improve readability?

pcgiang committed:
Updated Duke and TaskList to get correct response

over-fitted commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,190 @@
+    public Task markAsDone(String str) throws OutOfBoundException, NumberFormatException {
+        String i = str.substring(str.length()-1);
+        int index = Integer.parseInt(i);
+        if (index < 0 || index > list.size()) {
+            throw new OutOfBoundException();
+        } else {
+            Task task = list.get(index - 1);
+            task.markAsDone();
+            System.out.println("Nice! I have marked this task as done!");
+            System.out.println(task);
+            return task;
+        }
```
Could this logic block be refactored using a guard clause to handle the exception throwing in order to reduce indentation caused by the else block?

over-fitted commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,190 @@
+            //check if task if valid
+            if (str.startsWith("todo") || str.startsWith("deadline") || str.startsWith("event")) {
+                throw new EmptyDescriptionException(str);
+            } else { //check for invalid task
+                throw new InvalidTaskException();
+            }
```
Could this else be removed?

over-fitted commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,190 @@
+
+        //task that contain more than 1 word.
+        //check for invalid task. otherwise, add the correct task to the list.
+        else {
+            if (!str.startsWith("todo") && !str.startsWith("deadline") && !str.startsWith("event")) {
+                throw new InvalidTaskException();
+            }
+            else {
+                Task task;
+                if (str.startsWith("todo")) {
+                    task = new ToDos(str.substring(5));
+                    list.add(task);
+                } else if (str.startsWith("deadline")) {
+                    String[] message = str.split("/by ");
+                    if (message.length == 1) {
+                        throw new InvalidTaskException();
+                    } else {
+                        task = new Deadline(message[0].substring(9), message[1]);
+                        list.add(task);
+                    }
+                } else {
+                    String[] message = str.split("/at ");
+                    if (message.length == 1) {
+                        throw new InvalidTaskException();
+                    } else {
+                        task = new Events(message[0].substring(6), message[1]);
+                        list.add(task);
+                    }
+                }
+
+                System.out.println("Got it. I've added this task.");
+                System.out.println(list.get(list.size() - 1));
+                System.out.println("Now you have " + list.size()
+                        + (list.size() == 1 ? " task in the list" : " tasks in the list."));
+                return task;
+            }
+        }
```
Could the else statements be removed to improve readability?

over-fitted commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,190 @@
+
+        if (index < 0 || index > list.size()) {
+            throw new OutOfBoundException();
+        } else {
+            Task task = list.get(index - 1);
+            list.remove(index - 1);
+            System.out.println("Noted. I've removed this task: ");
+            System.out.println(task);
+            System.out.println("Now you have " + list.size()
+                    + (list.size() == 1 ? " task in the list" : " tasks in the list."));
+            return task;
+        }
```
since the if statement to throw the exception is a guard clause already, could the else statement be removed?

over-fitted reviewed (commented):
I like the comprehensiveness of your functional tests!

Just need to take another look at reducing the complexity in your code through reducing else block usage in multiple areas, in addition to a couple other flagged areas where complexity could be reduced. Further, it may prove useful for future maintainability to follow git commit title conventions. Otherwise, your PR for the most part LGTM! :tada:

pcgiang committed:
Removed extra print statement and cleaned up code

ananya-ravi-3021 commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,49 @@
+     */
+    public Deadline(String title, String deadline) throws InvalidDeadlineException {
+        super(title);
+        this.type = "D";
```
Perhaps you may not need the 'this' keyword here, since you are not assigning the instance variable 'type' to a parameter variable of the same name.

```suggestion
        type = "D";
```

ananya-ravi-3021 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,94 @@
+                        task = new Events(stuff[2], stuff[3]);
+                        break;
+                default:
+                        throw new InvalidTaskException();
```
Should there be a break statement here after throwing the InvalidTaskException?

```suggestion
                        throw new InvalidTaskException();
                        break
```

pcgiang replied:
I think there shouldn't be a break here since an exception is thrown? So the code wont be able to reach line 61 if im not wrong

ananya-ravi-3021 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,189 @@
+ *
+ */
+public class TaskList {
+    private ArrayList<Task> list;
```
Perhaps it would be better to name this variable tasks? Since it is a collection of Task objects.

```suggestion
    private ArrayList<Task> tasks;
```

ananya-ravi-3021 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,81 @@
+     * A method that allows user give input and runs the program.
+     * Contains a parser which reads command from the input.
+     */
+
```
Perhaps there should be no spaces between the documentation block and the method.

ananya-ravi-3021 commented on src/test/java/duke/test/DukeTest.java:
```diff
@@ -0,0 +1,15 @@
+package duke.test;
+import static org.junit.jupiter.api.Assertions.assertEquals;
```
Perhaps be more consistent with the ordering of imports, with static imports only above or below non-static imports.

ananya-ravi-3021 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,189 @@
+package duke;
+
+import duke.exception.*;
```
The asterisk in this import implies it is not explicit. Should it be changed to an explicit dukeException import?

ananya-ravi-3021 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,189 @@
+     * @throws NumberFormatException Thrown when user gives a non-number as the task number.
+     */
+    public String markAsDone(String str) throws OutOfBoundException, NumberFormatException {
+        String i = str.substring(str.length()-1);
```
Should the operator be surrounded with a space?
```suggestion
        String i = str.substring(str.length() - 1);
```

ananya-ravi-3021 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,189 @@
+    }
+
+    /**
+     * A method to delete a certain in a TaskList.
```
Perhaps an empty line between the parameter/return/throws section and the description?

Banana3021 reviewed (commented):
Overall, your program structure was excellent. The code is very easy to read and understand, and your documentation is very detailed. I found a small amount of coding standard violations, such as Javadoc comment spacing, and needing to include a 'break'  in the default component of the case statement.

Otherwise, really good job :)

pcgiang committed:
Add Assertions to highlight important assumptions

Assertions are not included in any methods.

Using assertions makes sure assumptions are verified during run time, hence avoid potential bugs.

Let's insert assertions for methods that involves type checking such as TaskList.

pcgiang committed:
Refractor code for better readability

Several methods do not follow indentation convention.
Some code fragments are not easy to understand.

Better indentation and if-else block makes code look clearer and improves readability.

Let's:
- remove unnecessary else block
- create new variables to break down complicated lines

pcgiang committed:
Merge pull request #1 from pcgiang/branch-A-CodeQuality

Refractor code for better readability

pcgiang committed:
Merge branch 'master' into branch-A-Assertions

pcgiang committed:
Merge pull request #2 from pcgiang/branch-A-Assertions

Branch a assertions

pcgiang committed:
Update Deadline and Event to include time

pcgiang committed:
Implement Undo. Refractor methods in TaskList to Command classes.

TaskList contains all methods and commands, therefore it is difficult
to keep track of sequence of command and how to undo each command.

Let's create Command abstract class and subclasses and implement undo inside each subclass.

pcgiang committed:
Implement Undo multiple steps.

Undo command only works for the last task.

Undo many steps would be useful for users.

Let's implement undo multiple steps by:
- Keep track of command history
- Create UndoableCommand class
- Update command history when a command is undone

pcgiang committed:
Update UndoCommand and check for invalid event input

pcgiang committed:
Finalise feature and UI. Add User Guide.

pcgiang committed:
Set theme jekyll-theme-minimal

pcgiang committed:
Update README.md

pcgiang committed:
Update level-7.txt

pcgiang committed:
Update UI.png

pcgiang committed:
Update Ui

pcgiang committed:
Update Ui
